### PR TITLE
KT-19428 Add inspection for redundant overrides that only call the super method

### PR DESCRIPTION
**Pull request by takahirom**
https://youtrack.jetbrains.com/issue/KT-19428

### CODE REVIEW

File: idea/src/org/jetbrains/kotlin/idea/inspections/KotlinRedundantOverrideInspection.kt [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+                    bodyExpression ?: return
+
+                    val body = if (bodyExpression is KtDotQualifiedExpression) {
+                        bodyExpression
+                    }
+                    else {
+                        if (bodyExpression.children.size != 1) {
+                            return
+                        }
+                        bodyExpression.children[0]
```
Comment (cypressious) [inline idea/src/org/jetbrains/kotlin/idea/inspections/KotlinRedundantOverrideInspection.kt]: You could use `singleOrNull() ?: return` here.

Comment (takahirom) [inline idea/src/org/jetbrains/kotlin/idea/inspections/KotlinRedundantOverrideInspection.kt]: I don't know it! Thanks :)

File: idea/src/org/jetbrains/kotlin/idea/inspections/KotlinRedundantOverrideInspection.kt [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+                    else {
+                        body as? KtDotQualifiedExpression ?: return
+                    }
+
+                    if (qualifiedExpression.receiverExpression !is KtSuperExpression) return
+
+                    val superSelectorExpression = qualifiedExpression.selectorExpression
+                    val superCallElement = superSelectorExpression as? KtCallElement ?: return
+
+                    if (!isSameFunctionName(superSelectorExpression, function)) return
```
Comment (takahirom) [inline idea/src/org/jetbrains/kotlin/idea/inspections/KotlinRedundantOverrideInspection.kt]: I can use `superCallElement` for the argument. I will fix it.

File: idea/testData/inspectionsLocal/redundantOverride/arguments.kt [hunk 3]
```diff
@@ -0,0 +1,10 @@
+open class Foo {
+    open fun arguments(arg1: Int, arg2: Long) {
+    }
+}
+
+class Bar : Foo() {
+    override <caret>fun arguments(arg1: Int, arg2: Long) {
```
Comment (mglukhikh) [inline idea/testData/inspectionsLocal/redundantOverride/arguments.kt]: I should mention that overridden function is not redundant if it has some other modifiers. For example, sometimes is reasonable to add `final` here, so function itself is not changed, but derived classes are made unable to override it. Also, I think that any annotation here makes the function not redundant. The same for any visibility modifier when you make overridden function more visible then original one. Please consider these cases, fix inspection, and add tests for them.

**Comment by mglukhikh** [comment]
Also, please add test when `super` function is called in expression body format:

```
override fun foo() = super.foo()
```

File: idea/src/org/jetbrains/kotlin/idea/inspections/KotlinRedundantOverrideInspection.kt [hunk 4]
```diff
@@ -69,6 +68,15 @@ class KotlinRedundantOverrideInspection : AbstractKotlinInspection(), CleanupLoc
                 }
             }
 
+    private fun KtNamedFunction.containsOtherModifier(): Boolean {
+        MODIFIER_EXCLUDE_OVERRIDE.forEach {
```
Comment (cypressious) [inline idea/src/org/jetbrains/kotlin/idea/inspections/KotlinRedundantOverrideInspection.kt]: You can simplify this to `return MODIFIER_EXCLUDE_OVERRIDE.any { hasModifier(it) }`

**Comment by takahirom** [comment]
@mglukhikh 
Done!

```
Also, please add test when super function is called in expression body format:

override fun foo() = super.foo()
```

Probably test of expression body format is already contained.
https://github.com/JetBrains/kotlin/pull/1242/files#diff-9a733327a84bcf25fd5047bae854a384R7

File: idea/src/org/jetbrains/kotlin/idea/inspections/KotlinRedundantOverrideInspection.kt (hunk 4 above)
Comment (takahirom) [inline idea/src/org/jetbrains/kotlin/idea/inspections/KotlinRedundantOverrideInspection.kt]: Thanks! I fixed it.

**Comment by takahirom** [comment]
I fixed conflicts.

**Comment by mglukhikh** [comment]
I think mostly it's Ok, I will recheck it, slightly correct style, and merge.

File: idea/src/org/jetbrains/kotlin/idea/inspections/KotlinRedundantOverrideInspection.kt [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean, session: LocalInspectionToolSession) =
+            object : KtVisitorVoid() {
+                override fun visitNamedFunction(function: KtNamedFunction) {
+                    super.visitNamedFunction(function)
+                    if (!function.hasModifier(KtTokens.OVERRIDE_KEYWORD)) return
+                    if (function.containsOtherModifier()) return
+                    if (function.annotationEntries.size != 0) return
+
+                    val bodyExpression = function.bodyExpression
+                    bodyExpression ?: return
```
Comment (mglukhikh) [inline idea/src/org/jetbrains/kotlin/idea/inspections/KotlinRedundantOverrideInspection.kt]: This and previous can be written in one line

File: idea/src/org/jetbrains/kotlin/idea/inspections/KotlinRedundantOverrideInspection.kt [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+import org.jetbrains.kotlin.psi.*
+import org.jetbrains.kotlin.psi.psiUtil.getCallNameExpression
+
+class KotlinRedundantOverrideInspection : AbstractKotlinInspection(), CleanupLocalInspectionTool {
+    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean, session: LocalInspectionToolSession) =
+            object : KtVisitorVoid() {
+                override fun visitNamedFunction(function: KtNamedFunction) {
+                    super.visitNamedFunction(function)
+                    if (!function.hasModifier(KtTokens.OVERRIDE_KEYWORD)) return
+                    if (function.containsOtherModifier()) return
```
Comment (mglukhikh) [inline idea/src/org/jetbrains/kotlin/idea/inspections/KotlinRedundantOverrideInspection.kt]: I'd inline this function

File: idea/src/org/jetbrains/kotlin/idea/inspections/KotlinRedundantOverrideInspection.kt [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+                    if (function.annotationEntries.size != 0) return
```
Comment (mglukhikh) [inline idea/src/org/jetbrains/kotlin/idea/inspections/KotlinRedundantOverrideInspection.kt]: isNotEmpty() ?

File: idea/src/org/jetbrains/kotlin/idea/inspections/KotlinRedundantOverrideInspection.kt [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+                                           RedundantOverrideFix())
+                }
+            }
+
+    private fun KtNamedFunction.containsOtherModifier(): Boolean = MODIFIER_EXCLUDE_OVERRIDE.any { hasModifier(it) }
+
+    private fun isSameArguments(superCallElement: KtCallElement, function: KtNamedFunction): Boolean {
+        val argumentList = superCallElement.valueArgumentList ?: return false
+        val parameterList = function.valueParameterList ?: return false
+        val diffArguments = argumentList.arguments.filter { argument ->
```
Comment (mglukhikh) [inline idea/src/org/jetbrains/kotlin/idea/inspections/KotlinRedundantOverrideInspection.kt]: This can be rewritten shorter using e.g. `none`

Comment (mglukhikh) [inline idea/src/org/jetbrains/kotlin/idea/inspections/KotlinRedundantOverrideInspection.kt]: Also, seems to me this logic works incorrectly if arguments are replaced.

**Comment by mglukhikh** [comment]
Merged manually

**Comment by takahirom** [comment]
Thank you for comments. I can learn it.

**Pull request closed without merging**